--- a/src/drivers.h
+++ b/src/drivers.h
@@ -160,6 +160,9 @@
 extern SensorDriver iio_buffer_compass;
 extern SensorDriver iio_poll_proximity;
 extern SensorDriver input_proximity;
+extern SensorDriver mtk_hf_accel;
+extern SensorDriver mtk_hf_light;
+extern SensorDriver mtk_hf_proximity;
 
 #ifdef HAS_LIBSSC
 extern SensorDriver ssc_proximity;
--- a/src/meson.build
+++ b/src/meson.build
@@ -24,6 +24,7 @@
 sources = [
   'iio-sensor-proxy.c',
   'drivers.c',
+  'drv-mtk-hf.c',
   'orientation.c',
   'drv-iio-buffer-accel.c',
   'drv-iio-poll-accel.c',
--- a/src/iio-sensor-proxy.c
+++ b/src/iio-sensor-proxy.c
@@ -68,6 +68,9 @@
 } SensorData;
 
 static const SensorDriver * const drivers[] = {
+        &mtk_hf_accel,
+        &mtk_hf_light,
+        &mtk_hf_proximity,
 	&iio_buffer_accel,
 	&iio_poll_accel,
 	&input_accel,
@@ -134,6 +137,7 @@
 	platform = g_udev_client_query_by_subsystem (client, "platform");
 	devices = g_list_concat (devices, input);
 	devices = g_list_concat (devices, platform);
+        devices = g_list_concat (devices, g_udev_client_query_by_subsystem (client, "hf_manager"));
 #ifdef HAS_LIBSSC
 	GList *fastrpc;
 	fastrpc = g_udev_client_query_by_subsystem (client, "misc");
@@ -716,6 +720,7 @@
 		"iio",
 		"input",
 		"platform",
+                "hf_manager",
 #ifdef HAS_LIBSSC
 		"misc",
 #endif
