diff --git a/arch/arm64/boot/dts/mediatek/mt6878.dtsi b/arch/arm64/boot/dts/mediatek/mt6878.dtsi
--- a/arch/arm64/boot/dts/mediatek/mt6878.dtsi
+++ b/arch/arm64/boot/dts/mediatek/mt6878.dtsi
@@ -61,8 +61,7 @@
 			reg = <0x0>;
 			performance-domains = <&performance 0>;
 			enable-method = "psci";
-			cpu-idle-states = <&cpuoff_l &clusteroff_l &mcusysoff_l
-						&system_vcore &s2idle>;
+			cpu-idle-states = <&cpuoff_l>;
 			capacity-dmips-mhz  = <267>;
 		};
 
@@ -72,8 +71,7 @@
 			reg = <0x100>;
 			performance-domains = <&performance 0>;
 			enable-method = "psci";
-			cpu-idle-states = <&cpuoff_l &clusteroff_l &mcusysoff_l
-						&system_vcore &s2idle>;
+			cpu-idle-states = <&cpuoff_l>;
 			capacity-dmips-mhz  = <267>;
 		};
 
@@ -83,8 +81,7 @@
 			reg = <0x200>;
 			performance-domains = <&performance 0>;
 			enable-method = "psci";
-			cpu-idle-states = <&cpuoff_l &clusteroff_l &mcusysoff_l
-						&system_vcore &s2idle>;
+			cpu-idle-states = <&cpuoff_l>;
 			capacity-dmips-mhz  = <267>;
 		};
 
@@ -94,8 +91,7 @@
 			reg = <0x300>;
 			performance-domains = <&performance 0>;
 			enable-method = "psci";
-			cpu-idle-states = <&cpuoff_l &clusteroff_l &mcusysoff_l
-						&system_vcore &s2idle>;
+			cpu-idle-states = <&cpuoff_l>;
 			capacity-dmips-mhz  = <267>;
 		};
 
@@ -105,8 +101,7 @@
 			reg = <0x400>;
 			performance-domains = <&performance 1>;
 			enable-method = "psci";
-			cpu-idle-states = <&cpuoff_b &clusteroff_b &mcusysoff_b
-						&system_vcore &s2idle>;
+			cpu-idle-states = <&cpuoff_b>;
 			capacity-dmips-mhz  = <1024>;
 		};
 
@@ -116,8 +111,7 @@
 			reg = <0x500>;
 			performance-domains = <&performance 1>;
 			enable-method = "psci";
-			cpu-idle-states = <&cpuoff_b &clusteroff_b &mcusysoff_b
-						&system_vcore &s2idle>;
+			cpu-idle-states = <&cpuoff_b>;
 			capacity-dmips-mhz  = <1024>;
 		};
 
@@ -127,8 +121,7 @@
 			reg = <0x600>;
 			performance-domains = <&performance 1>;
 			enable-method = "psci";
-			cpu-idle-states = <&cpuoff_b &clusteroff_b &mcusysoff_b
-						&system_vcore &s2idle>;
+			cpu-idle-states = <&cpuoff_b>;
 			capacity-dmips-mhz  = <1024>;
 		};
 
@@ -138,8 +131,7 @@
 			reg = <0x700>;
 			performance-domains = <&performance 1>;
 			enable-method = "psci";
-			cpu-idle-states = <&cpuoff_b &clusteroff_b &mcusysoff_b
-						&system_vcore &s2idle>;
+			cpu-idle-states = <&cpuoff_b>;
 			capacity-dmips-mhz  = <1024>;
 		};

