--- a/arch/arm64/boot/dts/mediatek/mt6878-nothing-tetris.dts
+++ b/arch/arm64/boot/dts/mediatek/mt6878-nothing-tetris.dts
@@ -5,6 +5,7 @@
 #include "mt6878.dtsi"
 #include <dt-bindings/gpio/gpio.h>
 #include <dt-bindings/input/input.h>
+#include <dt-bindings/usb/pd.h>
 
 / {
 	model = "Nothing CMF Phone 1";
@@ -127,7 +128,38 @@
 		status = "disabled";
 	};
 };
 
+&i2c4 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	clock-frequency = <400000>;
+	pinctrl-0 = <&rt1711h_int_default>;
+	pinctrl-names = "default";
+	status = "okay";
+
+	typec@4e {
+		compatible = "richtek,rt1711h";
+		reg = <0x4e>;
+		interrupt-parent = <&pio>;
+		interrupts = <41 IRQ_TYPE_LEVEL_LOW>;
+		wakeup-source;
+		status = "disabled";
+
+		connector {
+			compatible = "usb-c-connector";
+			label = "USB-C";
+			data-role = "dual";
+			power-role = "dual";
+			try-power-role = "sink";
+			source-pdos = <PDO_FIXED(5000, 500,
+				 PDO_FIXED_DUAL_ROLE | PDO_FIXED_DATA_SWAP)>;
+			sink-pdos = <PDO_FIXED(5000, 2000,
+				PDO_FIXED_DUAL_ROLE | PDO_FIXED_DATA_SWAP)>;
+			op-sink-microwatt = <10000000>;
+		};
+	};
+};
+
 &i2c5 {
 	#address-cells = <1>;
 	#size-cells = <0>;
@@ -105,6 +141,14 @@
 		};
 	};
 
+	rt1711h_int_default: rt1711h-int-default {
+		pins-int {
+			pinmux = <PINMUX_GPIO41__FUNC_GPIO41>;
+			input-enable;
+			bias-pull-up;
+		};
+	};
+
 	touch_pins_default: touch-default {
 		pins-i2c {
 			pinmux = <PINMUX_GPIO129__FUNC_SCL2>,
