diff --git a/arch/arm64/boot/dts/mediatek/mt6878.dtsi b/arch/arm64/boot/dts/mediatek/mt6878.dtsi
--- a/arch/arm64/boot/dts/mediatek/mt6878.dtsi
+++ b/arch/arm64/boot/dts/mediatek/mt6878.dtsi
@@ -61,7 +61,7 @@
 			reg = <0x0>;
 			performance-domains = <&performance 0>;
 			enable-method = "psci";
-			cpu-idle-states = <&cpuoff_l>;
+			cpu-idle-states = <&cpuoff_l &clusteroff_l>;
 			capacity-dmips-mhz  = <267>;
 		};
 
@@ -71,7 +71,7 @@
 			reg = <0x100>;
 			performance-domains = <&performance 0>;
 			enable-method = "psci";
-			cpu-idle-states = <&cpuoff_l>;
+			cpu-idle-states = <&cpuoff_l &clusteroff_l>;
 			capacity-dmips-mhz  = <267>;
 		};
 
@@ -81,7 +81,7 @@
 			reg = <0x200>;
 			performance-domains = <&performance 0>;
 			enable-method = "psci";
-			cpu-idle-states = <&cpuoff_l>;
+			cpu-idle-states = <&cpuoff_l &clusteroff_l>;
 			capacity-dmips-mhz  = <267>;
 		};
 
@@ -91,7 +91,7 @@
 			reg = <0x300>;
 			performance-domains = <&performance 0>;
 			enable-method = "psci";
-			cpu-idle-states = <&cpuoff_l>;
+			cpu-idle-states = <&cpuoff_l &clusteroff_l>;
 			capacity-dmips-mhz  = <267>;
 		};
 
@@ -101,7 +101,7 @@
 			reg = <0x400>;
 			performance-domains = <&performance 1>;
 			enable-method = "psci";
-			cpu-idle-states = <&cpuoff_b>;
+			cpu-idle-states = <&cpuoff_b &clusteroff_b>;
 			capacity-dmips-mhz  = <1024>;
 		};
 
@@ -111,7 +111,7 @@
 			reg = <0x500>;
 			performance-domains = <&performance 1>;
 			enable-method = "psci";
-			cpu-idle-states = <&cpuoff_b>;
+			cpu-idle-states = <&cpuoff_b &clusteroff_b>;
 			capacity-dmips-mhz  = <1024>;
 		};
 
@@ -121,7 +121,7 @@
 			reg = <0x600>;
 			performance-domains = <&performance 1>;
 			enable-method = "psci";
-			cpu-idle-states = <&cpuoff_b>;
+			cpu-idle-states = <&cpuoff_b &clusteroff_b>;
 			capacity-dmips-mhz  = <1024>;
 		};
 
@@ -131,7 +131,7 @@
 			reg = <0x700>;
 			performance-domains = <&performance 1>;
 			enable-method = "psci";
-			cpu-idle-states = <&cpuoff_b>;
+			cpu-idle-states = <&cpuoff_b &clusteroff_b>;
 			capacity-dmips-mhz  = <1024>;
 		};
 
