diff --git a/main/postmarketos-initramfs/00-initramfs-extra-base.files b/main/postmarketos-initramfs/00-initramfs-extra-base.files
index 396667ce6..d2bbae598 100644
--- a/main/postmarketos-initramfs/00-initramfs-extra-base.files
+++ b/main/postmarketos-initramfs/00-initramfs-extra-base.files
@@ -9,6 +9,7 @@
 /usr/lib/udev/rules.d/70-mouse.rules
 /usr/lib/udev/rules.d/70-touchpad.rules
 /usr/lib/udev/rules.d/80-drivers.rules
+/usr/lib/udev/rules.d/95-dm-notify.rules
 /usr/lib/udev/scsi_id
 /usr/lib/udev/udev.conf
 /sbin/btrfs
diff --git a/main/postmarketos-initramfs/APKBUILD b/main/postmarketos-initramfs/APKBUILD
index 1310b7aaa..985d39c81 100644
--- a/main/postmarketos-initramfs/APKBUILD
+++ b/main/postmarketos-initramfs/APKBUILD
@@ -1,13 +1,13 @@
 # Co-Maintainer: Clayton Craft <clayton@craftyguy.net>
 maintainer="Casey Connolly <kcxt@postmarketos.org>"
 pkgname=postmarketos-initramfs
-pkgver=3.12.0
-pkgrel=0
+pkgver=3.12.3
+pkgrel=1
 pkgdesc="Base files for the postmarketOS initramfs / initramfs-extra"
 url="https://postmarketos.org"
-options="!check"  # no tests
 provides="postmarketos-ramdisk=$pkgver-r$pkgrel"
 provider_priority=10
+checkdepends="postmarketos-testlib"
 depends="
 	beebzzr
 	blkid
@@ -50,6 +50,8 @@ source="
 	init_2nd.sh
 	init_functions_2nd.sh
 	mdev.conf
+	tests/01-parse-cmdline-testlib.sh
+	tests/02-find-partition-testlib.sh
 	"
 arch="noarch"
 license="GPL-2.0-or-later"
@@ -62,6 +64,10 @@ build() {
 		"$srcdir/init.sh" > "$builddir/init.sh"
 }
 
+check() {
+	postmarketos-testlib "$srcdir"
+}
+
 package() {
 	install -Dm644 "$srcdir"/mdev.conf \
 		-t "$pkgdir"/usr/share/initramfs/
@@ -110,10 +116,12 @@ sha512sums="
 ace852ba267b3408998970df6042603c8bec1d033066d1c30ed65f1be955fee8644b57882898f10228ec43b912f8f03718b34ddd741b431e5a5c2e76f060a494  00-default-extra.modules
 9c0e8f6f61d5da191e03a1aa9d5d0ceb5baf1eae6dbb9bfb0af59817783525119ac8394b135f303f7b6434a3eab0b49185fb90379e06823db847a4999c75ce33  00-initramfs-base.dirs
 69852928464f01757e2f473bd2c2e7ecb4132ecf3e26bdec95932e928dfe9042f59316e0a509fd38bcda88e24dd94bf33e8a88118b5adbec7023d1c13da12f59  00-initramfs-base.files
-6d91270b96ce0c745391174201dde63b43a73ae6456c2e1138046beef3548580588fedbeae7cc914079162e99343b45d82c6e6a7ee1d03829aae7697ae52f575  00-initramfs-extra-base.files
+669240c0c6b84ff68ea7696b3ffd39c209c7c6d52233f67a2515ac423e16c3daa7841d8451946b14dc95e1d18ff31cc35c4a22efb8a36caceb48aada5d4ee5ea  00-initramfs-extra-base.files
 bd72e8b6345a46ec13516c64d78d4fa1a9a18a8db9ab04aac605f9708629453dc65b6fbccd96b58eeb1ea68102ddbcae9dc6b369255c6c6c1fb18ce0bd43ec79  init.sh
-1c1518e903a1d08fe6eed85534fe84d12f43bb5fbfbca97de8aed78f30ddf96ebc2391d39eab02d8689549ca1367f422182995a26e5fda284afea024c433b1b9  init_functions.sh
+9765cad4f6c420a77dc975532e28c20fd89f9c6645b1c42e2fee98e3616729d30785e93affd7fef012a0118dd92d4e8b90718488bbd7782aa14424acd37d3e99  init_functions.sh
 7aa82d5e2907470d667612643d98371dff27b2650051f8c568ab43d1b97677362c3e6fbd442474086d681ea7c0cfbcaf269d3eef2d2dc3ff4ebf17c786b82148  init_2nd.sh
-4eca94327274511011967cde1f663b00b1a8182f7dddc3ba19d5c1f0ac094087a154a250c140ca70ac75a2e7c5bb70b03e94fa5eca63a5393efc4a23d025e1c3  init_functions_2nd.sh
+59f1e8954483dfe49182ddbaf0c1a4741f147161b92b5bb836e5bd405b419438f5757915546b46b846193c5a4b44c4b1a523f0ad0dcb87a1dfcaed9360e03146  init_functions_2nd.sh
 675e7d5bee39b2df7d322117f8dcaccc274d61beaf4d50ead19bbf2109446d64b1c0aa0c5b4f9846eb6c1c403418f28f6364eff4537ba41120fbfcbc484b7da7  mdev.conf
+776ac842df70a5ac7b20b942abf2afecfed83b59bd5bf775b77c1e8e071e1d9b9120c502410cd2643551589e9523221be0ca7a2734b2c5302fca7d631aa91916  01-parse-cmdline-testlib.sh
+a60b8affe7431d477f7480ca094721930f65a31f920c7eb937384ad708f98a4a4ab2ad68688bf92ff741d3083ce4f14c238ee2e1beae5cefc30b5fd9bfd6aa66  02-find-partition-testlib.sh
 "
diff --git a/main/postmarketos-initramfs/init_functions.sh b/main/postmarketos-initramfs/init_functions.sh
index c267d6228..936af83ce 100644
--- a/main/postmarketos-initramfs/init_functions.sh
+++ b/main/postmarketos-initramfs/init_functions.sh
@@ -105,6 +105,10 @@ parse_cmdline_item() {
 			# shellcheck disable=SC2034
 			log_info=y
 			;;
+		cryptkey)
+			# shellcheck disable=SC2034
+			cryptkey="$value"
+			;;
 		[![:alpha:]_]* | [[:alpha:]_]*[![:alnum:]_]*)
 			# invalid shell variable, ignore it
 			;;
@@ -909,6 +913,11 @@ start_unudhcpd() {
 	[ "$(pidof unudhcpd)" ] && return
 
 	local usb_iface
+	# Don't run if there's no USB gadget.
+	if [ -z "$(cat "$CONFIGFS/g1/UDC" 2>/dev/null)" ] &&
+		! [ -e /sys/class/android_usb/android0 ]; then
+		return
+	fi
 
 	# Skip if disabled
 	# shellcheck disable=SC2154
diff --git a/main/postmarketos-initramfs/init_functions_2nd.sh b/main/postmarketos-initramfs/init_functions_2nd.sh
index ab53e3edb..03c0f6e13 100644
--- a/main/postmarketos-initramfs/init_functions_2nd.sh
+++ b/main/postmarketos-initramfs/init_functions_2nd.sh
@@ -98,6 +98,9 @@ resize_root_partition() {
 unlock_root_partition() {
 	command -v cryptsetup >/dev/null || return
 	if cryptsetup isLuks "$PMOS_ROOT"; then
+		if [ -f "$cryptkey" ]; then
+			cryptsetup luksOpen "$PMOS_ROOT" root --key-file="$cryptkey"
+		fi
 		splash_hide
 		tried=0
 		until cryptsetup status root | grep -qwi active; do
diff --git a/main/postmarketos-initramfs/tests/.gitlab-ci.yml b/main/postmarketos-initramfs/tests/.gitlab-ci.yml
deleted file mode 100644
index 966f95adb..000000000
--- a/main/postmarketos-initramfs/tests/.gitlab-ci.yml
+++ /dev/null
@@ -1,9 +0,0 @@
-
-initramfs-unit-tests:
-  extends: .package-test
-  script:
-    - .ci/lib/testlib.sh main/postmarketos-initramfs/tests
-  rules:
-    - if: $CI_PIPELINE_SOURCE == "merge_request_event" || $CI_COMMIT_REF_PROTECTED == "false"
-      changes:
-        - main/postmarketos-initramfs/*.sh
diff --git a/main/postmarketos-initramfs/tests/01-parse-cmdline.sh b/main/postmarketos-initramfs/tests/01-parse-cmdline-testlib.sh
similarity index 95%
rename from main/postmarketos-initramfs/tests/01-parse-cmdline.sh
rename to main/postmarketos-initramfs/tests/01-parse-cmdline-testlib.sh
index 9f8b6005f..5ae5b8a50 100755
--- a/main/postmarketos-initramfs/tests/01-parse-cmdline.sh
+++ b/main/postmarketos-initramfs/tests/01-parse-cmdline-testlib.sh
@@ -21,7 +21,7 @@ start_test "Ensure all boolean variables can be set"
 # shellcheck disable=SC3001
 parse_cmdline < <(echo \
 pmos.bootchart2 pmos.debug-shell pmos.force-partition-resize \
-splash pmos.stowaway rd.info \
+splash pmos.stowaway rd.info cryptkey=/keyfile \
 )
 
 assert_equal "bootchart2" "y"
@@ -29,6 +29,7 @@ assert_equal "debug_shell" "y"
 assert_equal "force_partition_resize" "y"
 assert_equal "nosplash" "n"
 assert_equal "log_info" "y"
+assert_equal "cryptkey" "/keyfile"
 end_test
 
 ### Test 2 ###
diff --git a/main/postmarketos-initramfs/tests/02-find-partition.sh b/main/postmarketos-initramfs/tests/02-find-partition-testlib.sh
similarity index 100%
rename from main/postmarketos-initramfs/tests/02-find-partition.sh
rename to main/postmarketos-initramfs/tests/02-find-partition-testlib.sh
